### DISCUSSION TITLE
Detail Trackball Parts

### DISCUSSION THREAD

**Discussion by nethermead** [joshreve/dactyl-keyboard#41] [Ideas]
I've been playing with the trackball builds and I really like the direction it's all going.  (In general, this repo is terrific, thanks for it and for the work you're doing. If you need help, I'd be happy to lend a hand.)

I did get misdirected by another trackball build guide and bought some 3mm x 18mm steel rods for bearings that I blithely assumed would fit, but they're much too long for this model.  After poking around, I found a RollerBearingDowel.stl on the Ploopy Classic Trackball github and printed out three vertically and they seem to work nicely.  Note that all the STLs there seem to be scaled 10X larger, in cm instead of mm.

https://github.com/ploopyco/classic-trackball/tree/master/hardware/Mechanicals/STLs

I've attached a scaled STL.
[trackball_bearing_pin.zip](https://github.com/joshreve/dactyl-keyboard/files/7112218/trackball_bearing_pin.zip)

But would be good, when the trackball stuff is ready for primetime, to specify the bits and pieces needed to make it go: bearings, rods, 34mm ball, sensor, controller wiring, QMK funkiness, etc.  I'll be tackling a lot of this in the next week and I'll keep a log of gotchas and solutions.

**Comment by joshreve** [comment]
Still working on the build, but based on the interest may update the readme early to clarify for people.  All the hardware needed matches the Ploopy build.  The bearings are [on amazon](https://smile.amazon.com/dp/B085R8DNXG/ref=cm_sw_em_r_mt_dp_HRN0X2H9HNYTXZTJ2A6E?_encoding=UTF8&psc=1) and the posts and post/bearing press fixtures are actually just simple 3d prints available in the Ploopy repo.  I should pull them into my repo and add attributions as needed.  

The QMK side I am looking at right now.  I got a promicro with the PMW 3360 firmware running to confirm my hardware build, but will likely need to create a new QMK keyboard to support it all.  I currently have most of the pieces to start a full test build and will update the docs based on my learnings.

I am also investigating KMK and implementing a Pi Pico version to get the OLED, trackball, and per key LED running.  It doesn't look like any controller that is compatible with QMK is currently capable of this.  This is obviously a longer term objective.

**Reply by itised** [reply]
The develop branch of QMK should support the teensy 4. Not sure if that board will do what you want, but it's pretty versatile.

**Reply by joshreve** [reply]
Interesting.  The 4.0 looks like it may still be short i/o pins, but the 4.1 looks like it can handle anything.  This also looks like it signals the beginning of ARM support which should open up the field.  I also don't love the structure of QMK, but definitely would start over just for that reason.  The only other matter is that I don't want to pay ~$40 more ($20 per side) for an M7 when an M0 would do.  In reality though, I would if it saves me a ton of dev time.

Thanks for the tip.  If they really are starting serious ARM support, then I may not need to get creative.  I think the more power ARM architecture is better both due to typically having much more i/o and for compute bandwidth when you start bringing OLED, per key RGB, trackball, analog sticks, etc. into play at the same time.

**Reply by itised** [reply]
The teensy boards are great - I've used dozens of them in various projects, from the teensy 1 all the way to the 4.1, and I've never had any issues. They're very capable boards.

I've been wanting to use qmk on ARM for a while, so I've been following the progress of the proton-c and teensy support. But the teensy seems way more capable, and is actually able to be purchased. I really want to look into its USB host capabilities; I'd love to be able to plug a mouse into the keyboard so that I can use layers to alter the mouse buttons.

I've got a bunch of components set up for prototyping, so I've been trying splits with 3 or 4 "sides", multiple OLEDs (with different addresses or the same address and a multiplexer), lots of rotary encoders, trackballs, joysticks, etc., and I can say that qmk on ARM with a split board has been rough. To the point that I'm now doing all my prototyping using a single microcontroller, and connecting the different "sides" with HDMI cables. That way I can just wire everything together as one big matrix, with no need to communicate between microcontrollers over serial or I2C (which is pretty buggy). The teensy 4.1 is great for this - lots of IO and no need to deal with qmk communication between controllers, and HDMI gives me more than enough wires for the matrix + i2c + serial.

**Reply by joshreve** [reply]
I'm considering the balance of QMK with Circuit Python.  It may seem like I am Python-centric but in this case I am looking for a good abstraction layer.  I would also go with something based on the Arduino library.  What I don't want to have to do is mess with it every time a new controller is considered.  The arduino library and the circuit python library make a whole pile of controllers look identical and make everyone's life easier.  QMK tries with ChibiOS, but a keyboard doesn't need the complexity of an RTOS and it doesn't support even close to the same range of HW as arduino or circuit python.

Anyway, I am going to start with QMK, just to get some of the HW running and then work on finding the best approach for ARM.  I am currently leaning towards KMK since it will work with any circuit python compatible controller, including teensy 4, Pi Pico, and adafruit boards.  It is just missing some of the additional HW support, though the libraries exist so it is more about integrating them into the existing project.  If anyone incorporated Arduino library support, that may shift me in that direction as it would be a more processor efficient.

**Reply by joshreve** [reply]
I've been spending some time reading up and I think QMK is going to be a quagmire of high and low level integration issues.  I am going to investigate moving to Circuit Python.  I'm going to look at KMK, but first may just make a hack functional FW to check it out.

**Reply by joshreve** [reply]
Well, circuit python is not the way to go.  It lacks any low level capabilities.  No timing control sub milliseconds which is needed for some of the HW control.  I didn't realize it was really meant for people just starting to learn to do micro controller work.  I'm looking at what possibilities exist with Arduino.  It looks like the only readily available framework with hardware abstraction for a wide range of controllers and low level control.  I looked at creating an Arduino driven QMK, but QMK has low level references showing up all over the place with no clear abstraction or driver layer.

Honestly a little disheartened.  I am astounded how awesome the functionality is with QMK with such an entangled architecture.  It does explain why they don't have more of the new MCUs running.  I'll have to consider if I'm going to go down this path or just roll with whatever I can do with default QMK and slightly more obscure controllers.

**Reply by itised** [reply]
That's too bad. I was hoping it would be a viable option. 

I wonder how much work it would be to port just the matrix scanning code from QMK to Arduino. 

I imagine it's all you'd need, since all the other hardware should be supported by Arduino.

**Reply by joshreve** [reply]
That is what I'm going to look at, but the code is pretty tangled.  It looks like I may be able to extract the TMK and possibly Quantum portions of the code and reconnect them to arduino functions, but the interface boundary is not very clear.  I am debating if that is easier or hard than rewriting a simplified version.  It should be a LOT less code when building off existing arduino libraries.

There is also [Kaleidescope ](https://github.com/keyboardio/Kaleidoscope) from keyboardio which may be a better base as it is already build on top of Arduino, it just lacks some of the extra features for OLED, trackball, etc.  I'm going to read through the libraries a bit and try to find the best path.  I should still be able to pull some pieces out of QMK to add some features, but right now, I think it is almost easier to reimplement than try to find all the boundary interfaces to extract the QMK core.

**Reply by itised** [reply]
There's also an existing Arduino library for the matrix scanning:

https://www.arduino.cc/reference/en/libraries/keypad/

I'm not sure how robust or extensible it is, but it might be another good starting point

**Comment by joshreve** [comment]
I think I'm going to start from that or Kaleidescope.  I'm just going to make it what I want.  I do this for fun, so even if it takes a while I'll enjoy it.  I think I'm going to do something almost unheard of for small projects and actually create the architecture diagrams or at least a strongly documented API at each layer so people know how to work with it.  To really shine, I want to create it so people can add stuff without needing to understand the whole stack.  I think that often gets lost as one project evolves into another and intent and vision shifts.  I also want to try out Papyrus as one of the few open source architecture tools. 

I have a little more to wrap up to get my build running and this and a flex PCB are queued up to be next.  I'll have to decide which makes more sense to do first.  Probably the firmware so I know what controllers to focus on for the PCB.
